src/main/java/com/linkedin/kmf/consumer/NewConsumer.java

```diff
@@ -15,33 +15,99 @@
+
+  public NewConsumer(String topic, Properties consumerProperties, AdminClient adminClient)
+      throws ExecutionException, InterruptedException {
+    LOGGER.info("{} is being instantiated in the constructor..", this.getClass().getSimpleName());
 
-  public NewConsumer(String topic, Properties consumerProperties) {
+    NewConsumerConfig newConsumerConfig = new NewConsumerConfig(consumerProperties);
+    _targetConsumerGroupId = newConsumerConfig.getString(NewConsumerConfig.TARGET_CONSUMER_GROUP_ID_CONFIG);
+
+    if (_targetConsumerGroupId != null) {
+      this.configureGroupId(consumerProperties, adminClient);
+    }
     _consumer = new KafkaConsumer<>(consumerProperties);
     _consumer.subscribe(Collections.singletonList(topic));
   }
 
+  /**
+   * https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala#L189
+   * The consumer group string hash is used for this modulo operation.
+   * https://docs.confluent.io/current/clients/producer.html
+   * The partitioner will hash the key with murmur2 algorithm
+   * and divide it by the number of partitions.
+   * The result is that the same key is always assigned to the same partition.
+   * https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java#L69
+   * @param groupId kafka consumer group ID
+   * @param consumerOffsetsTopicPartitions number of partitions in the __consumer_offsets topic.
+   * @return hashed integer which represents a number, the Math.abs value of which is the broker
+   * ID of the group coordinator, or the leader of the offsets topic partition.
+   */
+  protected int consumerGroupCoordinatorHasher(String groupId, int consumerOffsetsTopicPartitions) {
+    // TODO: use long hash = UUID.nameUUIDFromBytes(s.getBytes()).getMostSignificantBits() ?
+    // TODO: MessageDigest digest = MessageDigest.getInstance("SHA-256");
+    // TODO: byte[] hash = digest.digest(text.getBytes(StandardCharsets.UTF_8));
+
+    LOGGER.debug("Hashed and modulo output: {}", groupId.hashCode());
+    return Math.abs(groupId.hashCode()) % consumerOffsetsTopicPartitions;
+  }
+
+  /**
+   * hash(group.id) % (number of __consumer_offsets topic partitions).
+   * The partition's leader is the group coordinator
+   * Choose B s.t hash(A) % (number of __consumer_offsets topic partitions) == hash(B) % (number of __consumer_offsets topic partitions)
+   * @param consumerProperties persistent set of kafka consumer properties
+   */
+  protected void configureGroupId(Properties consumerProperties, AdminClient adminClient)
+      throws ExecutionException, InterruptedException {
+    if (_targetConsumerGroupId.equals("")) {
+      throw new IllegalArgumentException("The target consumer group identifier cannot be empty: " + _targetConsumerGroupId);
+    }
+
+    int numOffsetsTopicPartitions = adminClient.describeTopics(Collections.singleton(Topic.GROUP_METADATA_TOPIC_NAME))
```

Comment:
Is AdminClient needed since Consumer.partitionsFor() should also return the list of partitions for a topic?

Reply:
Either AdminClient or KafkaConsumer works.